File: frontend/scripts/crap/method-parser.ts
/**
 * TypeScript source → list of FunctionDescriptor (start/end line, cyclomatic complexity).
 *
 * Mirrors `examples/crap4java/src/crap4java/JavaMethodParser.java`'s
 * MethodScanner + ComplexityCounter. Cyclomatic complexity counts branch
 * nodes per the McCabe definition: each function body starts at 1 and
 * increments on every if/loop/case/catch/?:/&&/||/?? — i.e. every place
 * where control flow can fork.
 */

import ts from "typescript";

export interface FunctionDescriptor {
	complexity: number;
	endLine: number;
	name: string;
	startLine: number; // 1-indexed
}

export function parseFunctions(filePath: string, source: string): FunctionDescriptor[] {
	const sourceFile = ts.createSourceFile(
		filePath,
		source,
		ts.ScriptTarget.ESNext,
		/* setParentNodes */ true,
		filePath.endsWith(".tsx") ? ts.ScriptKind.TSX : ts.ScriptKind.TS
	);

	const out: FunctionDescriptor[] = [];
	visit(sourceFile, sourceFile, out, []);
	return out;
}

function visit(
	node: ts.Node,
	sf: ts.SourceFile,
	out: FunctionDescriptor[],
	stack: ReadonlyArray<string>
): void {
	if (isFunctionLike(node)) {
		const name = describeName(node, stack);
		const { line: startLine } = sf.getLineAndCharacterOfPosition(node.getStart(sf));
		const { line: endLine } = sf.getLineAndCharacterOfPosition(node.getEnd());
		const complexity = computeComplexity(node);
		out.push({
			name,
			startLine: startLine + 1,
			endLine: endLine + 1,
			complexity,
		});
		const childStack = [...stack, name];
		ts.forEachChild(node, (child) => visit(child, sf, out, childStack));
		return;
	}
	ts.forEachChild(node, (child) => visit(child, sf, out, stack));
}

function isFunctionLike(node: ts.Node): node is ts.FunctionLikeDeclaration {
	return (
		ts.isFunctionDeclaration(node) ||
		ts.isMethodDeclaration(node) ||
		ts.isFunctionExpression(node) ||
		ts.isArrowFunction(node) ||
		ts.isConstructorDeclaration(node) ||
		ts.isGetAccessor(node) ||
		ts.isSetAccessor(node)
	);
}

function describeName(node: ts.FunctionLikeDeclaration, stack: ReadonlyArray<string>): string {
	const parent = node.parent;
	let local = "(anonymous)";
	if (ts.isFunctionDeclaration(node) && node.name) local = node.name.text;
	else if (ts.isMethodDeclaration(node) && node.name && ts.isIdentifier(node.name))
		local = node.name.text;
	else if (ts.isConstructorDeclaration(node)) local = "constructor";
	else if (ts.isGetAccessor(node) && ts.isIdentifier(node.name)) local = `get ${node.name.text}`;
	else if (ts.isSetAccessor(node) && ts.isIdentifier(node.name)) local = `set ${node.name.text}`;
	else if (parent && ts.isVariableDeclaration(parent) && ts.isIdentifier(parent.name))
		local = parent.name.text;
	else if (parent && ts.isPropertyAssignment(parent) && ts.isIdentifier(parent.name))
		local = parent.name.text;
	else if (parent && ts.isPropertyDeclaration(parent) && ts.isIdentifier(parent.name))
		local = parent.name.text;
	return stack.length === 0 ? local : `${stack[stack.length - 1]}>${local}`;
}

/**
 * Cyclomatic complexity counter. Walks ONLY this function's body (not nested
 * function bodies — those get their own descriptor). McCabe definition:
 * base 1 + 1 per branching node.
 */
function computeComplexity(fn: ts.FunctionLikeDeclaration): number {
	let cc = 1;
	function walk(node: ts.Node): void {
		// Don't descend into nested functions — their complexity is reported
		// separately via the outer visitor.
		if (node !== fn && isFunctionLike(node)) return;

		if (
			ts.isIfStatement(node) ||
			ts.isForStatement(node) ||
			ts.isForInStatement(node) ||
			ts.isForOfStatement(node) ||
			ts.isWhileStatement(node) ||
			ts.isDoStatement(node) ||
			ts.isCatchClause(node) ||
			ts.isConditionalExpression(node) ||
			ts.isCaseClause(node)
		) {
			cc++;
		}

		// Logical operators that introduce a branch: && || ??
		if (ts.isBinaryExpression(node)) {
			const op = node.operatorToken.kind;
			if (
				op === ts.SyntaxKind.AmpersandAmpersandToken ||
				op === ts.SyntaxKind.BarBarToken ||
				op === ts.SyntaxKind.QuestionQuestionToken
			) {
				cc++;
			}
		}

		ts.forEachChild(node, walk);
	}
	walk(fn);
	return cc;
}

File: examples/winstt-electron/server/scripts/crap/method_parser.py
"""Python source -> list of FunctionDescriptor (start/end line, cyclomatic complexity).

Mirrors `frontend/scripts/crap/method-parser.ts` (the TS-AST visitor), using
the stdlib ``ast`` module instead of the TypeScript compiler API.

Cyclomatic complexity counts branch nodes per the McCabe definition: each
function body starts at 1 and increments on every place control flow can
fork. Nested functions/lambdas are NOT counted toward the enclosing
function — they get their own descriptor (same rule as the TS port).

Python-specific branch mapping (radon-aligned):

    if / elif / for / while / except / ternary (a if c else b)   -> +1 each
    match case (non-wildcard; ``case _`` is the `default:` analog) -> +1 each
    comprehension generator (implicit loop)                        -> +1 each
    each ``if`` filter inside a comprehension                      -> +1 each
    bool op chain ``a and b and c``                                -> +(n-1)

``with`` / ``try`` / ``assert`` are not branch points (McCabe; matches the TS
port which counts only catch clauses, not try blocks).
"""

from __future__ import annotations

import ast
from dataclasses import dataclass

FunctionLike = (ast.FunctionDef, ast.AsyncFunctionDef, ast.Lambda)


@dataclass(frozen=True)
class FunctionDescriptor:
    name: str
    start_line: int  # 1-indexed
    end_line: int
    complexity: int


def parse_functions(file_path: str, source: str) -> list[FunctionDescriptor]:
    try:
        tree = ast.parse(source, filename=file_path)
    except SyntaxError:
        return []
    out: list[FunctionDescriptor] = []
    _visit(tree, out, [])
    return out


def _local_name(node: ast.AST) -> str:
    if isinstance(node, ast.Lambda):
        return "<lambda>"
    if isinstance(node, (ast.FunctionDef, ast.AsyncFunctionDef, ast.ClassDef)):
        return node.name
    return ""


def _visit(node: ast.AST, out: list[FunctionDescriptor], stack: list[str]) -> None:
    is_fn = isinstance(node, FunctionLike)
    is_cls = isinstance(node, ast.ClassDef)

    if is_fn:
        # ``isinstance(node, FunctionLike)`` already narrowed ``node`` for runtime
        # but mypy needs the explicit assert to see ``.lineno`` / ``decorator_list``.
        assert isinstance(node, (ast.FunctionDef, ast.AsyncFunctionDef, ast.Lambda))
        local = _local_name(node)
        qualified = ".".join([*stack, local]) if stack else local
        start = node.lineno
        # Fold decorator lines into the span so decorator-line coverage counts.
        decorators: list[ast.expr] = getattr(node, "decorator_list", [])
        if decorators:
            start = min(start, min(d.lineno for d in decorators))
        end = getattr(node, "end_lineno", None) or start
        out.append(
            FunctionDescriptor(
                name=qualified,
                start_line=start,
                end_line=end,
                complexity=_compute_complexity(node),
            )
        )

    child_stack = [*stack, _local_name(node)] if (is_fn or is_cls) else stack
    for child in ast.iter_child_nodes(node):
        _visit(child, out, child_stack)


def _is_wildcard_case(case: ast.match_case) -> bool:
    """``case _:`` (and bare ``case x:`` capture) is the `default:` analog —
    the TS port does not count DefaultClause, so we skip the catch-all too."""
    pat = case.pattern
    return case.guard is None and isinstance(pat, ast.MatchAs) and pat.pattern is None


def _compute_complexity(fn: ast.FunctionDef | ast.AsyncFunctionDef | ast.Lambda) -> int:
    cc = 1

    def walk(node: ast.AST) -> None:
        nonlocal cc
        # Don't descend into nested functions/lambdas — reported separately.
        if node is not fn and isinstance(node, FunctionLike):
            return

        if isinstance(
            node,
            (ast.If, ast.For, ast.AsyncFor, ast.While, ast.ExceptHandler, ast.IfExp),
        ):
            cc += 1
        elif isinstance(node, ast.match_case):
            if not _is_wildcard_case(node):
                cc += 1
        elif isinstance(node, ast.comprehension):
            cc += 1  # the implicit loop
            cc += len(node.ifs)  # each filter predicate
        elif isinstance(node, ast.BoolOp):
            cc += len(node.values) - 1

        for child in ast.iter_child_nodes(node):
            walk(child)

    walk(fn)
    return cc
